hazelcast-client-new/src/main/java/com/hazelcast/client/spi/impl/discovery/DiscoveryAddressTranslator.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.client.spi.impl.discovery;
+
+import com.hazelcast.client.connection.AddressTranslator;
+import com.hazelcast.nio.Address;
+import com.hazelcast.spi.discovery.DiscoveredNode;
+import com.hazelcast.spi.discovery.integration.DiscoveryService;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class DiscoveryAddressTranslator
+        implements AddressTranslator {
+
+    private final DiscoveryService discoveryService;
+
+    private volatile Map<Address, Address> privateToPublic;
+
+    public DiscoveryAddressTranslator(DiscoveryService discoveryService) {
+        this.discoveryService = discoveryService;
+    }
+
+    @Override
+    public Address translate(Address address) {
+        if (address == null) {
+            return null;
+        }
+
+        // Refresh only once to prevent load on service discovery
+        boolean alreadyRefreshed = false;
```

Comment:
yeah discoverNodes() asks at the discovery provider, if that one doesn't cache (which mostly won't make sense) you refresh the assignment but you don't want to do it too often.